prov/efa/src/rxr/rxr_pkt_type_misc.c

```diff
@@ -63,45 +63,46 @@ ssize_t rxr_pkt_init_handshake(struct rxr_ep *ep,
 
 	pkt_entry->pkt_size = sizeof(struct rxr_handshake_hdr)
 			      + RXR_NUM_PROTOCOL_VERSION * sizeof(uint64_t);
-	pkt_entry->addr = addr;
+	pkt_entry->addr = rx_entry->addr;
+	pkt_entry->x_entry = rx_entry;
 	return 0;
 }
 
-void rxr_pkt_post_handshake(struct rxr_ep *ep,
-			    struct rxr_peer *peer,
-			    fi_addr_t addr)
+ssize_t rxr_pkt_post_handshake(struct rxr_ep *ep, struct rxr_peer *peer,
+				fi_addr_t addr)
 {
-	struct rxr_pkt_entry *pkt_entry;
+	struct rxr_rx_entry *rx_entry;
 	ssize_t ret;
 
 	assert(!(peer->flags & RXR_PEER_HANDSHAKE_SENT));
 
-	pkt_entry = rxr_pkt_entry_alloc(ep, ep->tx_pkt_efa_pool);
-	if (OFI_UNLIKELY(!pkt_entry))
-		return;
+	rx_entry = ofi_buf_alloc(ep->rx_entry_pool);
+	if (OFI_UNLIKELY(!rx_entry)) {
+		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL, "RX entries exhausted.\n");
+		return -FI_ENOMEM;
+	}
 
-	rxr_pkt_init_handshake(ep, pkt_entry, addr);
+	rx_entry->addr = addr;
+	dlist_init(&rx_entry->queued_pkts);
+	rx_entry->type = RXR_RX_ENTRY;
+	rx_entry->state = RXR_RX_INIT;
 
-	/*
-	 * TODO: Once we start using a core's selective completion capability,
-	 * post the HANDSHAKE packets without FI_COMPLETION.
-	 */
-	ret = rxr_pkt_entry_send(ep, pkt_entry, addr);
+#if ENABLE_DEBUG
+	dlist_insert_tail(&rx_entry->rx_entry_entry, &ep->rx_entry_list);
+#endif
 
-	/*
-	 * Skip sending this handshake on error and try again when processing the
-	 * next REQ from this peer containing the source information
-	 */
+	ret = rxr_pkt_post_ctrl_or_queue(ep, RXR_RX_ENTRY, rx_entry, RXR_HANDSHAKE_PKT, 0);
 	if (OFI_UNLIKELY(ret)) {
-		rxr_pkt_entry_release_tx(ep, pkt_entry);
-		if (ret == -FI_EAGAIN)
-			return;
-		FI_WARN(&rxr_prov, FI_LOG_CQ,
-			"Failed to send a HANDSHAKE packet: ret %zd\n", ret);
-		return;
+		// Once error happens, pkt_entry has been released within rxr_pkt_post_ctrl_once
+		rxr_release_rx_entry(ep, rx_entry);
+		FI_WARN(&rxr_prov, FI_LOG_EP_CTRL,
+			"Failed to send a HANDSHAKE packet: %s\n", fi_strerror(-ret));
+		return ret;
```

Comment:
Why not just call `efa_eq_write_error` here, and do not return error code. This way the caller does not need to change.